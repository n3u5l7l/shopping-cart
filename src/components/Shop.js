import "../styles/Shop.css";
import { motion, AnimatePresence } from "framer-motion";
import CashewPic from "../assets/cashew-nuts.jpeg";
import AlmondPic from "../assets/Almond.jpeg";
import WalnutPic from "../assets/Walnut.jpeg";
import PecanPic from "../assets/Pecans.jpeg";
import PistachiosPic from "../assets/Pistachios.jpeg";
import HazlenutPic from "../assets/Hazelnut.jpeg";
import MacadamiaNutPic from "../assets/Macadamia-nut.jpeg";
import PeanutPic from "../assets/peanut.jpeg";
import BrazilNutPic from "../assets/Brazil-nut.jpeg";

function Shop({ handleShoppingCart: setShoppingCart }) {
  const NutInfos = [
    {
      image: CashewPic,
      type: "Cashew nuts",
      cost: "15.99",
    },
    {
      image: AlmondPic,
      type: "Almonds",
      cost: "25.99",
    },
    {
      image: WalnutPic,
      type: "Walnuts",
      cost: "14.99",
    },
    {
      image: PecanPic,
      type: "Pecans",
      cost: "14.99",
    },
    {
      image: PistachiosPic,
      type: "Pistachios",
      cost: "20.99",
    },
    {
      image: HazlenutPic,
      type: "Hazelnuts",
      cost: "18.99",
    },
    {
      image: BrazilNutPic,
      type: "Brazil nuts",
      cost: "15.99",
    },
    {
      image: MacadamiaNutPic,
      type: "Macadamia nuts",
      cost: "10.99",
    },
    {
      image: PeanutPic,
      type: "Peanuts",
      cost: "17.99",
    },
  ];

  const updateCart = (nutType, nutCost, nutImg) => {
    setShoppingCart((prevState) =>
      Object.assign(
        {},
        prevState,
        prevState[nutType] && prevState[nutType].amount >= 0
          ? {
              [nutType]: {
                ...prevState[nutType],
                amount: Number(prevState[nutType].amount) + 1,
                total:
                  Number(prevState[nutType].amount) *
                  Number(prevState[nutType].cost),
              },
              totalCost: prevState["totalCost"] + nutCost,
              totalItem: prevState["totalItem"] + 1,
            }
          : {
              [nutType]: { amount: 1, cost: nutCost, image: nutImg },
              totalCost: prevState.totalCost
                ? prevState.totalCost + nutCost
                : nutCost,
              totalItem: prevState.totalItem ? prevState.totalItem + 1 : 1,
            }
      )
    );
    console.log("LoL");
  };
  let content = NutInfos.map((nut, index) => {
    return (
      <div className="item-container" key={index}>
        <img src={nut.image} alt={nut.type + " pic"} />
        <div className="item-info">
          <div className="nut-type">{nut.type}</div>
          <div className="nut-cost">${nut.cost}</div>
          <button
            onClick={() => updateCart(nut.type, Number(nut.cost), nut.image)}
          >
            Add to cart
          </button>
        </div>
      </div>
    );
  });

  return (
    <motion.div
      className="shop-container"
      initial={{ x: 300, opacity: 0 }}
      animate={{ x: 0, opacity: 1 }}
      exit={{ x: window.innerWidth, opacity: 1, transition: { duration: 0.3 } }}
    >
      <div className="shop-title">Nut Collection</div>
      <div className="all-item-container">{content}</div>
    </motion.div>
  );
}

export default Shop;