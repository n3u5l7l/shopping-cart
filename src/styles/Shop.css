#root{
    --grid-col: auto-fit;
}
.shop-container{
    padding:3rem 5rem 3rem 5rem;
    display: flex;
    flex-direction: column;
    margin-top:50px;
}
.shop-title{
    color:black;
    background-color: white;
    font-size: 1.6rem;
    display: flex;
    justify-content: center;
    font-weight: bolder;
    padding-top:2rem;
    border-bottom: 1px solid;
}
.all-item-container{
    display:grid;
    grid-template-columns: repeat(var(--grid-col, auto-fit), 250px);
    gap:50px;
    padding:30px;
    justify-content: center;
    background-color: white;
    padding:2rem;
}

@media (min-width: 1200px) {
    #root{
        --grid-col: auto-fit;
    }
}

.item-container{
    display:flex;
    flex-direction: column;
    gap:7px;
    box-shadow: 0 0 5px #ccc;
    transform: scale(1);
    transition: transform 0.3s ease-in-out;
}

.item-container:hover{
    transform: scale(1.1);
}

.item-container > img{
    width:100%;
    height:auto;
    flex:1;
}

.item-info{
    padding:10px;
    display: flex;
    flex-direction: column;
    gap:5px;
}

.nut-type{
    font-size: 1.2rem;
    font-weight: bolder;
}

.item-info > button{
    border:none;
    height:30px;
    opacity:0.7;
    background-color: black;
    color: white;
    cursor: pointer;
    transition: opacity 0.3s ease-in-out;
}

.item-info > button:hover{
    opacity:1;
}