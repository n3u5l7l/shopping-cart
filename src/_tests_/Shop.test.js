import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import Shop from "../components/Shop";
import React from "react";
import userEvent from "@testing-library/user-event";

afterEach(cleanup);

it("renders properly", () => {
    const {container} = render(<Shop />);

    expect(container).toMatchSnapshot();
});

it("renders all shop item button", async () => {
    render(<Shop />);
    const allButtons = screen.getAllByRole('button');

    expect(allButtons).toHaveLength(allButtons.length);
});

it("renders all images", () => {
    render(<Shop />);
    
    const allImages = screen.getAllByRole("img");

    expect(allImages).toHaveLength(allImages.length);
});

test("all buttons can be clicked on shop items", async () => {
    const updateCartMock = jest.fn();
    render(<Shop updateCart = {updateCartMock}/>);
    const allButtons = screen.getAllByRole('button');

    for(let i = 0; i < allButtons.length; i++){
        await userEvent.click(allButtons[i]);
    }

    expect(updateCartMock).toHaveBeenCalledTimes(allButtons.length);
});



