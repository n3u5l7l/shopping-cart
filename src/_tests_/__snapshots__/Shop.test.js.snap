// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div>
  <div
    class="shop-container"
    style="opacity: 0; transform: translateX(300px) translateZ(0);"
  >
    <div
      class="shop-title"
    >
      Nut Collection
    </div>
    <div
      class="all-nut-container"
    >
      <section
        class="nut-container"
      >
        <img
          alt="Cashew nuts pic"
          src="cashew-nuts.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Cashew nuts
          </div>
          <div
            class="nut-cost"
          >
            $
            15.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Almonds pic"
          src="Almond.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Almonds
          </div>
          <div
            class="nut-cost"
          >
            $
            25.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Walnuts pic"
          src="Walnut.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Walnuts
          </div>
          <div
            class="nut-cost"
          >
            $
            14.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Pecans pic"
          src="Pecans.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Pecans
          </div>
          <div
            class="nut-cost"
          >
            $
            14.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Pistachios pic"
          src="Pistachios.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Pistachios
          </div>
          <div
            class="nut-cost"
          >
            $
            20.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Hazelnuts pic"
          src="Hazelnut.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Hazelnuts
          </div>
          <div
            class="nut-cost"
          >
            $
            18.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Brazil nuts pic"
          src="Brazil-nut.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Brazil nuts
          </div>
          <div
            class="nut-cost"
          >
            $
            15.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Macadamia nuts pic"
          src="Macadamia-nut.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Macadamia nuts
          </div>
          <div
            class="nut-cost"
          >
            $
            10.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
      <section
        class="nut-container"
      >
        <img
          alt="Peanuts pic"
          src="peanut.jpeg"
        />
        <div
          class="nut-info"
        >
          <div
            class="nut-type"
          >
            Peanuts
          </div>
          <div
            class="nut-cost"
          >
            $
            17.99
          </div>
          <button>
            Add to cart
          </button>
        </div>
      </section>
    </div>
  </div>
</div>
`;
